what-the-stack/src/main/java/com/haroldadmin/whatthestack/WhatTheStackService.kt

```diff
@@ -20,15 +32,88 @@ import android.os.*
  * exception data as an intent extra.
  */
 @ServiceProcess
-class WhatTheStackService : Service() {
+class WhatTheStackService : LifecycleService() {
+    private var finishedWaitingForFile = false
+    private var stopAfterWaitingForFile = false
+
+    override fun onCreate() {
+        super.onCreate()
+
+        lifecycle.coroutineScope.launch {
+            val earlyStartupCrashFile = getStartupCrashFile(this@WhatTheStackService)
+            for (i in 0 until 10) {
+                if (earlyStartupCrashFile.exists()) {
+                    val crashDataRaw = earlyStartupCrashFile.readBytes()
+                    val parcel = Parcel.obtain()
+                    val crashBundle = try {
+                        parcel.unmarshall(crashDataRaw, 0, crashDataRaw.size)
+                        parcel.setDataPosition(0)
+                        Bundle.CREATOR.createFromParcel(parcel)
+                    } finally {
+                        parcel.recycle()
+                    }
+
+                    getStartupCrashFile(applicationContext).delete()
+
+                    // We cannot open the activity after app has already died because of
+                    // background activity start restrictions. Show notification instead.
+                    showCrashNotification(crashBundle)
+
+                    stopSelf()
+
+                    break
+                }
+                delay(500)
+            }
+
+            if (stopAfterWaitingForFile) {
+                stopSelf()
+            }
+            finishedWaitingForFile = true
+        }
+    }
+
+    private fun showCrashNotification(crashBundle: Bundle) {
```

Comment:
What happens if we don't have notification permissions? Do we just silently fail?
Not sure what the best approach there would be, but I know leak canary can throw up a permission dialog, which I'm not a big fan of

Reply:
We can't throw up any dialogs or show the user anything, because we are considered "in background" (foreground app has already crashed and closed). And since these types crashes happen on startup, we cannot show the user any other UI.

So when app starts crashing on startup, user must first go to the System settings, enable notifications for the app and then open the app again (to crash again to show the notification).

Do you think we should include the notification permission in the manifest just to be sure app add it?

it is a bit shitty workaround, but I don't see any other option. Any ideas are welcome.

Reply:
I assume that if we don't have notification permission that the service crashes when trying to send the notification?

I think it would be better if we can try to avoid pulutting the log with crashes from the library, as they can hide the original stacktrace and make it hard to check the logs

What if we add a check if we can send notification and we cannot, print a warning/error to the log?

Reply:
I've checked and it does not crash if there is no notification permission. But you are right that warning would be a good idea. Added that.